Kernel/Process.cpp

```diff
@@ -63,6 +63,67 @@ SpinlockProtected<Process::List>& Process::all_instances()
     return *s_all_instances;
 }
 
+ErrorOr<void> Process::for_each_with_jail_respect(Function<ErrorOr<void>(Process&)> callback)
+{
+    ErrorOr<void> result {};
+    jail().with([&](auto my_jail) {
+        Process::all_instances().with([&](auto const& list) {
+            for (auto it = list.begin(); it != list.end();) {
+                auto& process = *it;
+                ++it;
+                process.jail().with([&](auto& their_jail) {
+                    if (!my_jail || their_jail.ptr() == my_jail.ptr())
+                        result = callback(process);
+                });
+                if (result.is_error())
+                    break;
+            }
+        });
+    });
```

Comment:
This dance makes me nervous. We take the spinlocks in the order Process::jail --> Process::all_instances --> Process::jail. Is this always safe?

Is it also looks like it will lock `this`'s jail spinlock twice once we get to it in the all processes list?

I'm not convinced that iterating this list requires manual `it` varaibles the way you're doing it. does ranged for work?

Do we really want to call the callback while holding all 3 spin locks?

I wonder if it would be ok to aggregate a list of processes in a Vector with some inline capacity here on the stack and call the callback after gathering them under the lock?

Reply:
I changed it a bit now and it should take a different path to this. Also, if we keep taking our jail spinlock and the second process spinlock under the spinlock of `Process::all_instances()`, in theory this should prevent a deadlock from happening because no two processes can compete over each other jail spinlock unless they somehow both acquire the `Process::all_instances` which should never happen...